include/oneapi/internal/execution_context_sycl.h

```diff
@@ -41,164 +42,105 @@ namespace interface1
 {
 class OpenClKernelFactory : public Base, public ClKernelFactoryIface
 {
-private:
-    class ProgramCacheEntry
-    {
-    private:
-        OpenClProgramRef * _program;
+public:
+    explicit OpenClKernelFactory(cl::sycl::queue & deviceQueue) : _currentProgramRef(nullptr),
+        _executionTarget(ExecutionTargetIds::unspecified), _deviceQueue(deviceQueue) {}
 
-    public:
-        ProgramCacheEntry() : _program(nullptr) {}
+    ~OpenClKernelFactory() DAAL_C11_OVERRIDE {}
 
-        ~ProgramCacheEntry()
-        {
-            delete _program;
-        }
 
-        void setProgram(OpenClProgramRef *program)
-        {
-            _program = program;
-        }
-
-        OpenClProgramRef * getProgram()
+    void build(ExecutionTargetId target, const char * name, const char * program, const char * options = "",
+               services::Status * status = nullptr) DAAL_C11_OVERRIDE
+    {
+        // TODO: Thread safe?
+        services::Status local_status;
+        services::String key = name;
+        const bool res = programHashTable.contain(key, local_status);
+        if (!local_status.ok())
         {
-            return _program;
+            services::internal::tryAssignStatus(status, local_status);
+            return;
         }
-
-        const char* getName(services::Status * status = nullptr)
+        if (!res)
         {
-            if (!_program)
+            auto programPtr = services::SharedPtr<OpenClProgramRef>(new OpenClProgramRef(_deviceQueue.get_context().get(),
+                _deviceQueue.get_device().get(), name, program, options, &local_status));
+            if (!local_status.ok())
+            {
+                services::internal::tryAssignStatus(status, local_status);
+                return;
+            }
+            programHashTable.add(key, programPtr, local_status);
+            if (!local_status.ok())
             {
-                if (status)
-                {
-                    status->add(services::ErrorExecutionContext);
-                }
-                return nullptr;
+                services::internal::tryAssignStatus(status, local_status);
+                return;
             }
-            return _program->getName();
         }
-    };
-
-    class KernelCacheEntry
-    {
-    private:
-        KernelPtr _kernel;
-        services::String _name;
-
-    public:
-        KernelCacheEntry() : _kernel(), _name() {}
-
-        ~KernelCacheEntry() {}
 
-        void setKernel(KernelPtr kernel, const char *name)
+        _currentProgramRef = programHashTable.get(key, local_status).get();
```

Comment:
In case when you adding a program to cache and immediately requesting it, you can return it directly from `add()` method to avoid additional unnecessary logic overhead.